File: packages/template/src/surfaces/acp/session-manager.ts
/**
 * ACP Session Manager
 *
 * Tracks active ACP sessions and provides close/list operations that
 * deepagents-acp does not natively support. Pure in-memory bookkeeping; the
 * lifecycle patching that drives it lives in `session-lifecycle.ts`.
 */

import { logger } from "../../runtime/logger.js";

export interface SessionInfo {
  sessionId: string;
  createdAt: string;
  lastActivityAt: string;
  mode: string;
  messageCount: number;
}

/**
 * Session lifecycle manager. Tracks active sessions and provides
 * close/list operations that deepagents-acp doesn't natively support.
 */
export class SessionManager {
  private sessions = new Map<string, SessionInfo>();
  private log = logger.child("session-manager");

  track(sessionId: string, mode: string): void {
    this.sessions.set(sessionId, {
      sessionId,
      createdAt: new Date().toISOString(),
      lastActivityAt: new Date().toISOString(),
      mode,
      messageCount: 0,
    });
    this.log.debug("Session tracked", { sessionId, mode, total: this.sessions.size });
  }

  touch(sessionId: string): void {
    const info = this.sessions.get(sessionId);
    if (info) {
      info.lastActivityAt = new Date().toISOString();
      info.messageCount++;
    }
  }

  close(sessionId: string): SessionInfo | undefined {
    const info = this.sessions.get(sessionId);
    if (info) {
      this.sessions.delete(sessionId);
      this.log.info("Session closed", { sessionId, messages: info.messageCount });
    }
    return info;
  }

  list(): SessionInfo[] {
    return Array.from(this.sessions.values());
  }

  has(sessionId: string): boolean {
    return this.sessions.has(sessionId);
  }

  get count(): number {
    return this.sessions.size;
  }
}

File: packages/template-python/src/deepagents_template/surfaces/acp/session_manager.py
"""ACP Session Manager — tracks active ACP sessions."""

from __future__ import annotations

from datetime import UTC, datetime
from typing import Any


class SessionManager:
    def __init__(self) -> None:
        self._sessions: dict[str, dict[str, Any]] = {}

    def track(self, sessionId: str, mode: str) -> None:
        self._sessions[sessionId] = {
            "sessionId": sessionId,
            "createdAt": datetime.now(UTC).isoformat(),
            "lastActivityAt": datetime.now(UTC).isoformat(),
            "mode": mode,
            "messageCount": 0,
        }

    def touch(self, sessionId: str) -> None:
        info = self._sessions.get(sessionId)
        if info:
            info["lastActivityAt"] = datetime.now(UTC).isoformat()
            info["messageCount"] = info.get("messageCount", 0) + 1

    def close(self, sessionId: str) -> dict[str, Any] | None:
        return self._sessions.pop(sessionId, None)

    def list(self) -> list[dict[str, Any]]:
        return list(self._sessions.values())

    def has(self, sessionId: str) -> bool:
        return sessionId in self._sessions

    @property
    def count(self) -> int:
        return len(self._sessions)
